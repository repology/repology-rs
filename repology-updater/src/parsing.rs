// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

pub mod error;
pub mod package_maker;
pub mod parser;
pub mod parsers;
pub mod sink;
pub mod sinks;
pub mod utils;
