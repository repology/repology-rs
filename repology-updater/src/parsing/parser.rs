// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use crate::parsing::sink::PackageSink;

pub trait PackageParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()>;
}
