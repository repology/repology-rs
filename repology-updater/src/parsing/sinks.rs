// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::package::ParsedPackage;
use crate::parsing::error::PackageParsingError;
use crate::parsing::package_maker::PackageMaker;
use crate::parsing::sink::PackageSink;

#[derive(Default)]
pub struct PackageAccumulator {
    pub packages: Vec<ParsedPackage>,
}

impl PackageSink for PackageAccumulator {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError> {
        self.packages.push(package_maker.finalize()?);
        Ok(())
    }
}

#[derive(Default)]
pub struct PackageNullSink {}

impl PackageSink for PackageNullSink {
    fn push(&mut self, _: PackageMaker) -> Result<(), PackageParsingError> {
        Ok(())
    }
}

pub struct PackageCounter<'n, N: ?Sized> {
    pub count: usize,
    next: &'n mut N,
}

impl<'n, N> PackageCounter<'n, N>
where
    N: PackageSink + ?Sized,
{
    pub fn new(next: &'n mut N) -> Self {
        Self { count: 0, next }
    }
}

impl<'n, N> PackageSink for PackageCounter<'n, N>
where
    N: PackageSink + ?Sized,
{
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError> {
        self.count += 1;
        self.next.push(package_maker)
    }
}

#[derive(Default)]
pub struct PackageDumper {}

impl PackageSink for PackageDumper {
    fn push(&mut self, package_maker: PackageMaker) -> Result<(), PackageParsingError> {
        println!("{:#?}", package_maker.finalize()?);
        Ok(())
    }
}
