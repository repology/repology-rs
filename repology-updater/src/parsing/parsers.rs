// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

pub mod freebsd;
pub mod stalix;
pub mod tincan;
pub mod yacp;

pub use freebsd::*;
pub use stalix::*;
pub use tincan::*;
pub use yacp::*;

use crate::parsing::parser::PackageParser;

pub fn create_parser(name: &str) -> anyhow::Result<Box<dyn PackageParser>> {
    match name {
        "FreeBsdParser" => Ok(Box::new(freebsd::FreeBsdParser {})),
        "StalIxParser" => Ok(Box::new(stalix::StalIxParser {})),
        "TinCanParser" => Ok(Box::new(tincan::TinCanParser {})),
        "YacpParser" => Ok(Box::new(yacp::YacpParser {})),
        _ => Err(anyhow::anyhow!("invalid parser name {}", name)),
    }
}
