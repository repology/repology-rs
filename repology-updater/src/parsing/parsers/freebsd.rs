// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use anyhow::{Context, anyhow, bail};
use std::io::BufRead as _;
use std::path::Path;

use repology_common::LinkType;

use crate::parsing::package_maker::{NameType, PackageMaker};
use crate::parsing::parser::PackageParser;
use crate::parsing::sink::PackageSink;
use crate::parsing::utils::maintainers::extract_maintainers;
use crate::parsing::utils::version::VersionStripper;

const EXPECTED_FIELDS_COUNT: usize = 13;
const VERSION_STRIPPER: VersionStripper = VersionStripper::new()
    .with_strip_right(",")
    .with_strip_right("_");

pub struct FreeBsdParser {}

impl FreeBsdParser {
    fn parse_line(line: &str, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let fields: Vec<_> = line.trim().split('|').collect();

        if fields.len() != EXPECTED_FIELDS_COUNT {
            bail!(
                "expected {} fields, got {}",
                EXPECTED_FIELDS_COUNT,
                fields.len()
            );
        }

        let mut pkg = PackageMaker::default();

        let (package_name, version) = fields[0]
            .rsplit_once('-')
            .ok_or_else(|| anyhow!("expected <package name>-<version> in the first field"))?;

        let mut path_comps = fields[1].rsplit('/');
        let port_name = path_comps
            .next()
            .expect("split always returns at least one component");
        let port_category = path_comps.next().ok_or_else(|| {
            anyhow!("unexpectedly short port path (expected at least category/name)")
        })?;

        pkg.set_names(package_name, NameType::BinName | NameType::ProjectNameSeed);
        pkg.set_names(
            format!("{port_category}/{port_name}"),
            NameType::SrcName | NameType::DisplayName | NameType::TrackName,
        );
        pkg.set_version_stripped(version, &VERSION_STRIPPER);
        pkg.set_summary(fields[3]);
        pkg.add_maintainers(extract_maintainers(fields[5]));
        pkg.add_categories(fields[6].split_ascii_whitespace());
        pkg.add_links(
            LinkType::UpstreamHomepage,
            fields[9].split_ascii_whitespace(),
        );

        Ok(sink.push(pkg)?)
    }
}

impl PackageParser for FreeBsdParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let f = std::fs::File::open_buffered(path)?;

        for (nline, line) in f.lines().enumerate() {
            Self::parse_line(&line?, sink).with_context(|| format!("on line {}", nline + 1))?;
        }

        Ok(())
    }
}
