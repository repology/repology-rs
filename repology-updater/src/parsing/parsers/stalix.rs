// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::io::BufRead as _;
use std::path::Path;

use anyhow::{Context, bail};

use repology_common::LinkType;

use crate::parsing::package_maker::{NameType, PackageMaker};
use crate::parsing::parser::{PackageParser, PackageSink};
use crate::parsing::utils::maintainers::extract_maintainers;

#[allow(unused)]
mod data {
    use std::collections::HashMap;

    use serde::Deserialize;

    #[derive(Deserialize)]
    #[cfg_attr(feature = "strict-parsers", serde(deny_unknown_fields))]
    pub struct Package {
        pub category: String,
        pub ix_pkg_fs_name: String,
        pub ix_pkg_name: String,
        pub ix_pkg_full_name: String,
        pub pkg_name: String,
        pub pkg_ver: String,
        pub recipe: String,
        pub maintainers: Vec<String>,
        pub upstream_urls: Vec<String>,
        pub lang_module: Option<String>,
    }
}

pub struct StalIxParser {}

impl StalIxParser {
    fn parse_line(line: &str, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let pkgdata: data::Package = serde_json::from_str(line)?;

        // a hack because we have no way to pass arbitrary flag to the ruleset
        // in fact the prefix should be added by 500ths rules
        let prefix = match pkgdata.lang_module.as_deref() {
            Some("python") => "python:",
            Some("perl") => "perl:",
            None => "",
            Some(other) => {
                bail!("unexpected lang_module {other}");
            }
        };

        let mut pkg = PackageMaker::default();

        pkg.set_names(
            pkgdata.ix_pkg_full_name,
            NameType::SrcName | NameType::TrackName,
        );
        pkg.set_names(
            pkgdata.ix_pkg_name,
            NameType::BinName | NameType::DisplayName,
        );
        pkg.set_names(
            format!("{prefix}{}", pkgdata.pkg_name),
            NameType::ProjectNameSeed,
        );

        pkg.set_version(pkgdata.pkg_ver);
        pkg.add_category(pkgdata.category);
        for maintainer in &pkgdata.maintainers {
            pkg.add_maintainers(extract_maintainers(maintainer));
        }
        for url in &pkgdata.upstream_urls {
            pkg.add_link(LinkType::UpstreamDownload, url);
        }

        Ok(sink.push(pkg)?)
    }
}

impl PackageParser for StalIxParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let f = std::fs::File::open_buffered(path)?;

        for (nline, line) in f.lines().enumerate() {
            Self::parse_line(&line?, sink).with_context(|| format!("on line {}", nline + 1))?;
        }

        Ok(())
    }
}
