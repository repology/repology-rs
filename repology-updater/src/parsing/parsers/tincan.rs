// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use anyhow::{Context, anyhow, bail};

use repology_common::LinkType;

use crate::parsing::package_maker::{NameType, PackageMaker};
use crate::parsing::parser::PackageParser;
use crate::parsing::sink::PackageSink;
use crate::parsing::utils::maintainers::extract_maintainers;
use crate::parsing::utils::walk::{WalkEntry, WalkFileTree};

#[allow(unused)]
mod data {
    use std::collections::HashMap;

    use serde::Deserialize;

    #[derive(Deserialize)]
    #[cfg_attr(feature = "strict-parsers", serde(deny_unknown_fields))]
    pub struct Meta {
        pub version: String,
        pub maintainer: String,
        pub sources: Vec<String>,
        pub checksums: Vec<String>,
    }

    #[derive(Deserialize)]
    #[cfg_attr(feature = "strict-parsers", serde(deny_unknown_fields))]
    pub struct Package {
        pub meta: Meta,
        pub deps: HashMap<String, String>,
        pub mkdeps: HashMap<String, String>,
    }
}

pub struct TinCanParser {}

impl TinCanParser {
    fn process_package(entry: &WalkEntry, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let package_path_absolute = entry
            .path_absolute
            .parent()
            .expect("should be able to get parent path of package.toml");
        let package_subdir = package_path_absolute
            .file_name()
            .and_then(|file_name| file_name.to_str())
            .ok_or_else(|| anyhow!("cannot extract name of package subdirectory"))?;

        let pkgdata: data::Package =
            toml::from_str(std::str::from_utf8(&std::fs::read(&entry.path_absolute)?)?)?;

        let mut pkg = PackageMaker::default();

        pkg.set_names(
            package_subdir,
            NameType::SrcName
                | NameType::TrackName
                | NameType::DisplayName
                | NameType::ProjectNameSeed,
        );
        pkg.set_version(pkgdata.meta.version);
        pkg.add_maintainers(extract_maintainers(&pkgdata.meta.maintainer));

        let mut patches = vec![];
        for source in &pkgdata.meta.sources {
            if source.contains("://") {
                pkg.add_link(LinkType::UpstreamDownload, source);
            } else if source.starts_with("files/") && source.ends_with(".patch") {
                if !std::fs::exists(package_path_absolute.join(source))? {
                    bail!(
                        "patch file {source} mentioned in meta.sources section was not found on the file system"
                    );
                } else {
                    patches.push(source);
                }
            }
        }

        pkg.set_extra_field_many("patches", patches);

        Ok(sink.push(pkg)?)
    }
}

impl PackageParser for TinCanParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        for entry in WalkFileTree::walk_by_name(path, "package.toml") {
            let entry = entry?;
            Self::process_package(&entry, sink)
                .with_context(|| format!("while processing {:?}", entry.path_relative))?;
        }

        Ok(())
    }
}
