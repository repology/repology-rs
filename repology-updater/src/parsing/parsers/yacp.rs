// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use anyhow::Context;

use repology_common::LinkType;

use crate::parsing::package_maker::{NameType, PackageMaker};
use crate::parsing::parser::PackageParser;
use crate::parsing::sink::PackageSink;

#[allow(unused)]
mod data {
    use serde::Deserialize;

    #[derive(Deserialize)]
    #[cfg_attr(feature = "strict-parsers", serde(deny_unknown_fields))]
    pub struct SubPackage {
        pub name: String,
        pub category: Vec<String>,
    }

    #[derive(Deserialize)]
    #[cfg_attr(feature = "strict-parsers", serde(deny_unknown_fields))]
    pub struct Package {
        pub name: String,
        pub version: String,
        pub category: Vec<String>,
        pub summary: String,
        pub homepage: String,
        pub subpackages: Vec<SubPackage>,
        pub maintainers: Vec<String>,
    }

    #[derive(Deserialize)]
    #[cfg_attr(feature = "strict-parsers", serde(deny_unknown_fields))]
    pub struct Data {
        pub repository_name: String,
        pub num_packages: usize,
        pub timestamp: u64,
        pub packages: Vec<Package>,
    }
}

pub struct YacpParser {}

impl YacpParser {
    fn process_package(package: data::Package, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let mut pkg = PackageMaker::default();

        pkg.set_names(
            package.name,
            NameType::SrcName
                | NameType::TrackName
                | NameType::DisplayName
                | NameType::ProjectNameSeed,
        );
        pkg.set_version(package.version);
        pkg.add_categories(package.category);
        pkg.set_summary(package.summary);
        pkg.add_link(LinkType::UpstreamHomepage, package.homepage);
        pkg.add_binnames(
            package
                .subpackages
                .into_iter()
                .map(|subpackage| subpackage.name),
        );

        Ok(sink.push(pkg)?)
    }
}

impl PackageParser for YacpParser {
    fn parse(&self, path: &Path, sink: &mut dyn PackageSink) -> anyhow::Result<()> {
        let data: data::Data = serde_json::from_reader(std::fs::File::open_buffered(path)?)?;

        for (npackage, package) in data.packages.into_iter().enumerate() {
            Self::process_package(package, sink)
                .with_context(|| format!("entry #{}", npackage + 1))?;
        }

        Ok(())
    }
}
