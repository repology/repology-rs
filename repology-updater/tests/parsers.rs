// SPDX-FileCopyrightText: Copyright 2025 Dmitry Marakasov <amdmi3@amdmi3.ru>
// SPDX-License-Identifier: GPL-3.0-or-later

use std::path::Path;

use paste::paste;

use repology_updater::parsing::parser::PackageParser as _;
use repology_updater::parsing::parsers::*;
use repology_updater::parsing::sinks::PackageAccumulator;

macro_rules! parser_test {
    ($parser_name:ident, $test_name:ident) => {
        paste! {
            #[test]
            fn [<test_ $parser_name:lower _ $test_name>]() {
                let mut packages = PackageAccumulator::default();
                let res = [<$parser_name Parser>] {}
                    .parse(Path::new(concat!("tests/parsers_test_data/", stringify!([<$parser_name:lower>]), "/", stringify!($test_name))), &mut packages).map(|_| packages.packages);
                insta::assert_debug_snapshot!(res);
            }
        }
    };
}

parser_test!(FreeBsd, ok);
parser_test!(FreeBsd, error_missing_field);
parser_test!(FreeBsd, error_extra_field);
parser_test!(FreeBsd, error_missing_version);
parser_test!(FreeBsd, error_missing_name);
parser_test!(FreeBsd, error_bad_package_format);
parser_test!(FreeBsd, error_bad_package_path);

parser_test!(TinCan, ok);

parser_test!(Yacp, ok);

parser_test!(StalIx, ok);
parser_test!(StalIx, error_bad_lang_module);
parser_test!(StalIx, error_bad_json);
