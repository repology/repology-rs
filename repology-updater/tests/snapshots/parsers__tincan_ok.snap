---
source: repology-updater/tests/parsers.rs
expression: res
snapshot_kind: text
---
Ok(
    [
        Package {
            srcname: "busybox",
            trackname: "busybox",
            visiblename: "busybox",
            projectname_seed: "busybox",
            rawversion: "1.36.1",
            maintainers: [
                "avs.origami@gmail.com",
            ],
            links: [
                Link {
                    type: UpstreamDownload,
                    url: "https://busybox.net/downloads/busybox-1.36.1.tar.bz2",
                    fragment: None,
                },
            ],
            version: "1.36.1",
            flags: PackageFlags(
                0x0,
            ),
        },
        Package {
            srcname: "libelf",
            trackname: "libelf",
            visiblename: "libelf",
            projectname_seed: "libelf",
            rawversion: "0.191",
            maintainers: [
                "avs.origami@gmail.com",
            ],
            extrafields: {
                "patches": ManyValues(
                    [
                        "files/musl.patch",
                    ],
                ),
            },
            links: [
                Link {
                    type: UpstreamDownload,
                    url: "https://sourceware.org/elfutils/ftp/0.191/elfutils-0.191.tar.bz2",
                    fragment: None,
                },
            ],
            version: "0.191",
            flags: PackageFlags(
                0x0,
            ),
        },
        Package {
            srcname: "musl",
            trackname: "musl",
            visiblename: "musl",
            projectname_seed: "musl",
            rawversion: "1.2.5",
            maintainers: [
                "avs.origami@gmail.com",
            ],
            links: [
                Link {
                    type: UpstreamDownload,
                    url: "https://musl.libc.org/releases/musl-1.2.5.tar.gz",
                    fragment: None,
                },
            ],
            version: "1.2.5",
            flags: PackageFlags(
                0x0,
            ),
        },
        Package {
            srcname: "rustup",
            trackname: "rustup",
            visiblename: "rustup",
            projectname_seed: "rustup",
            rawversion: "1.28.2",
            maintainers: [
                "avs.origami@gmail.com",
            ],
            links: [
                Link {
                    type: UpstreamDownload,
                    url: "https://static.rust-lang.org/rustup/archive/1.28.2/x86_64-unknown-linux-musl/rustup-init",
                    fragment: None,
                },
            ],
            version: "1.28.2",
            flags: PackageFlags(
                0x0,
            ),
        },
    ],
)
